server/app/services/rpc_server.rb

```diff
@@ -13,30 +22,81 @@ def initialize(code, message, backtrace = nil)
     end
   end
 
-  ##
+  attr_reader :handlers
+
+  def initialize
+    @handlers = {}
+  end
+
+  # @param [Faye::Websocket] ws_client
   # @param [String] grid_id
   # @param [Array] message msgpack-rpc request array
   # @return [Array]
-  def self.handle_request(grid_id, message)
+  def handle_request(ws_client, grid_id, message)
     msg_id = message[1]
-    grid = Grid.find(grid_id)
-    return [1, msg_id, {error: 'grid not found'}, nil] if grid.nil?
-
     handler = message[2].split('/')[1]
     method = message[2].split('/')[2]
-    if klass = HANDLERS[handler]
+    if actor = handling_actor(grid_id, handler)
       begin
-        result = klass.new(grid).send(method, *message[3])
-        return [1, msg_id, nil, result]
+        result = actor.send(method, *message[3])
+        send_message(ws_client, [1, msg_id, nil, result])
+      rescue Celluloid::DeadActorError
+        error "actor for handler #{handler} is dead, removing it from cache"
+        @handlers[grid_id].delete(handler)
       rescue RpcServer::Error => exc
-        return [1, msg_id, {code: exc.code, message: exc.message, backtrace: exc.backtrace}, nil]
+        send_message(ws_client, [1, msg_id, {code: exc.code, message: exc.message, backtrace: exc.backtrace}, nil])
+        @handlers[grid_id].delete(handler)
       rescue => exc
-        puts "#{exc.class.name}: #{exc.message}"
-        puts exc.backtrace.join("\n")
-        return [1, msg_id, {code: 500, message: "#{exc.class.name}: #{exc.message}", backtrace: exc.backtrace}, nil]
+        error "#{exc.class.name}: #{exc.message}"
+        debug exc.backtrace.join("\n")
+        send_message(ws_client, [1, msg_id, {code: 500, message: "#{exc.class.name}: #{exc.message}", backtrace: exc.backtrace}, nil])
+        @handlers[grid_id].delete(handler)
       end
     else
-      return [1, msg_id, {error: 'service not implemented'}, nil]
+      warn "handler #{handler} not implemented"
+      send_message(ws_client, [1, msg_id, {code: 501, error: 'service not implemented'}, nil])
     end
   end
+
+  # @param [String] grid_id
+  # @param [Array] message msgpack-rpc notification array
+  def handle_notification(grid_id, message)
+    handler = message[1].split('/')[1]
+    method = message[1].split('/')[2]
+    if actor = handling_actor(grid_id, handler)
+      begin
+        debug "rpc notification: #{actor.class.name}##{method} #{message[2]}"
```

Comment:
This makes the server `LOG_LEVEL=0` significantly more verbose than before... but also more useful. It's just pretty spammy now with all the very large parameter hashes being logged.